### ISSUE TITLE
Unable to boot newly created image

### ISSUE THREAD

**Issue by cwoolum** [sgielen/picl-k3os-image-generator#22]
Hello, I used the docker image to create a Raspberry PI Image. Upon trying to boot, I get a `Kernel panic - not syncing: Requested init /sbin/init.resizefs failed (error -2)`.

I checked the troubleshooting steps and thought the issue may have been with the SD Card so I tried flashing another card and have the same issue. 

I am on Windows and build the image in Docker. I then used Balena etcher to write the image to the card.

I did mount the volume to try and debug `/sbin/init.resizefs` but that file does not appear to exist.

Here are my logs from docker run

<details>
<summary>Click to expand</summary>

``` sh

Building an image for the Raspberry Pi model 3B+/4.
== Checking or downloading dependencies... ==
RASPBERRY_PI_FIRMWARE env variable was not set - defaulting to known good firmware [1.20200212]
== Making image and filesystems... ==
sh: 1: udevadm: not found
sh: 1: udevadm: not found
sh: 1: udevadm: not found
sh: 1: udevadm: not found
sh: 1: udevadm: not found
sh: 1: udevadm: not found
sh: 1: udevadm: not found
sh: 1: udevadm: not found
sh: 1: udevadm: not found
sh: 1: udevadm: not found
/dev/loop3: msdos partitions 1 2
mkfs.fat 4.1 (2017-01-24)
mke2fs 1.44.1 (24-Mar-2018)
Creating filesystem with 195072 4k blocks and 48768 inodes
Filesystem UUID: c2dee55b-048c-428a-857c-1cde613c7e95
Superblock backups stored on blocks:
        32768, 98304, 163840

Allocating group tables: done
Writing inode tables: done
Creating journal (4096 blocks): done
Writing superblocks and filesystem accounting information: done

tune2fs 1.44.1 (24-Mar-2018)
Setting interval between checks to 2592000 seconds
== Initializing root... ==
== Initializing boot... ==
== Installing... ==

== picl-k3os-v0.10.0-raspberrypi.img created. ==
```
</details>

**Comment by sgielen** [comment]
Error -2 is ENOENT, indicating it can't find /sbin/init.resizefs. This file only exists on first boot: it resizes the root volume, sets the proper init in cmdline, then deletes itself and its dependencies.

Do you have screen output when your Raspberry Pi boots? The first time it boots, the output should contain something like "Resizing root filesystem on..." - subsequent boots don't have this line in the output and immediately boot into k3os.

Could you write the image to your SD card again, and see if you can get the output of the first boot? If that works fine, but the second boot fails, something is going wrong with updating the cmdline. In that case, I'm interested in the contents of your /boot/cmdline.txt file.

If the first boot already gives the error you indicate, could you write the image freshly to the SD card and inspect of /sbin/init.resizefs does exist then?

**Comment by cwoolum** [comment]
Here's the screenshot of first boot.

![20200729_124101](https://user-images.githubusercontent.com/942415/88845564-36ee2680-d199-11ea-8899-9f7684e17c23.jpg)

I flashed a fresh card and mounted it and `/sbin/init.resizefs` still seems to be missing.

I tried creating the image without docker as well. The udevadm warnings go away but I still get the error.

**Comment by cwoolum** [comment]
This is my `init.resizefs`

<details>
<summary>Click here to open</summary>

``` sh
#!/bin/sh

set -e

IMAGE_TYPE="raspberrypi"

# sleep a second for devices to settle
sleep 1

mount -t proc none /proc
mount -t sysfs none /sys

# Unpack tools for resizing root FS
mount -t tmpfs -o size=50m none /tmp
tar -xJf /root-resize.tar.xz --strip 1 -C /tmp

export PATH=/sbin:/bin:/usr/sbin:/usr/bin:/tmp/sbin:/tmp/bin:/tmp/usr/sbin:/tmp/usr/bin
export LD_LIBRARY_PATH=/tmp/lib:/tmp/lib/aarch64-linux-gnu:/tmp/usr/lib:/tmp/usr/lib/aarch64-linux-gnu

if [ "$IMAGE_TYPE" = "raspberrypi" ]; then
	PARTUUID=$(cat /proc/cmdline | sed -r 's/^.*PARTUUID=([^ ]+).*$/\1/')
	ROOTDEVICE=$(blkid | grep PARTUUID=\"$PARTUUID\" | awk -F: '{print $1}')
elif [ "$IMAGE_TYPE" = "orangepipc2" ]; then
	ROOTDEVICE=$(cat /proc/cmdline | sed -r 's/.*root=([^ ]+) .*/\1/')
fi
ROOTDISK=/dev/$(basename $(dirname $(readlink /sys/class/block/$(basename $ROOTDEVICE))))
PARTNUM=$(cat /sys/class/block/$(basename $ROOTDEVICE)/partition)

echo "== Performing filesystem check on root device $ROOTDEVICE... =="
e2fsck -pv $ROOTDEVICE
sleep 5
echo "== Resizing root filesystem on $ROOTDISK... =="
parted -s $ROOTDISK resizepart $PARTNUM 100%
sleep 5
mount -o remount,rw /
resize2fs $ROOTDEVICE
sleep 5

echo "== Setting proper init... =="
if [ "$IMAGE_TYPE" = "raspberrypi" ]; then
	mount ${ROOTDISK}p1 /boot
	sed -i 's# init=/sbin/init.resizefs# init=/sbin/init.preinit#' /boot/cmdline.txt
	NEW_PARTUUID=$(blkid -o export $ROOTDEVICE | grep PARTUUID)
	sed -i "s# root=PARTUUID=$PARTUUID # root=$NEW_PARTUUID #" /boot/cmdline.txt
	umount /boot
elif [ "$IMAGE_TYPE" = "orangepipc2" ]; then
	sed -i 's# init=/sbin/init.resizefs# init=/sbin/init.preinit#' /boot/env.txt
fi

echo "== Linking correct config... =="
if [ "$IMAGE_TYPE" = "orangepipc2" ]; then
	modprobe dwmac_sun8i
	sleep 1
fi

# TODO: what if this device doesn't have eth0, but has other interfaces? Loop through them?
MY_MAC=$(cat /sys/class/net/eth0/address)
if [ -r "/k3os/system/config/${MY_MAC}.yaml" ]; then
	ln -s /k3os/system/config/${MY_MAC}.yaml /k3os/system/config.yaml
else
	# TODO: retrieve config from a master?
	echo "Error: No configuration found for MAC ${MY_MAC}." >&2
	echo "Ensure that config/${MY_MAC}.yaml exists." >&2
	exit 1
fi

echo "== Cleaning up and rebooting... =="
rm /root-resize.tar.xz
rm /sbin/init.resizefs
# exec into another script so that /sbin/init.resizefs (this script) can be deleted,
# so the filesystem can be mounted read-only
cat <<EOF >/tmp/reboot
#!/bin/sh
set -e
sync
sleep 5
mount -o remount,ro / || echo u > /proc/sysrq-trigger || true
sleep 5
e2fsck -fyv $ROOTDEVICE || true
sleep 5
reboot -f
EOF
chmod +x /tmp/reboot
exec /tmp/reboot

```

</details>

**Comment by sgielen** [comment]
So, if I understand correctly, after you create an image, flash it to a USB drive and then immediately mount the partition (without booting the Raspberry Pi with it), you don't have a file at `/sbin/init.resizefs`?

Just to be sure: you are flashing the image `picl-k3os-v0.10.0-raspberrypi.img` and its last modified date is approximately when the Docker run process finished?

Could you check whether you have a file at `/sbin/init.preinit` on the flashed partition? Also, could you check if you see a `/root-resize.tar.xz` file?

These files are copied in build-image.sh around line 340:

```
sudo tar -cJf root/root-resize.tar.xz "root-resize"
sudo rm -rf root-resize

## Write a resizing init and a pre-init
sudo install -m 0755 -o root -g root init.preinit init.resizefs root/sbin                                                                                 
sudo sed -i "s#@IMAGE_TYPE@#$IMAGE_TYPE#" root/sbin/init.resizefs root/sbin/init.preinit
```

I wonder if you put a `ls -la root/sbin/init*` there, what the resulting output is.

**Comment by sgielen** [comment]
Oh, wait, I re-read your comment now and I understand that `/sbin/init.resizefs` does exist on the SD card, but your boot still fails with "Requested init /sbin/init.resizefs failed (error -2)."

That only makes slightly more sense, though, let me look into that.

**Comment by sgielen** [comment]
Could you modify your `/boot/cmdline.txt` to change `init=/sbin/init.resizefs` to `init=/bin/sh`? When you boot this, it should spawn a shell that allows you to interact. It'll probably just give you a "# " prompt.

Once you get that, could you run `ls -la /sbin/init.resizefs` to see what the state of that file is? Then, after that, could you run `exec /sbin/init.resizefs` to see what happens?

**Comment by sgielen** [comment]
Since you're the first person reporting you're building the image on Windows, I'm suspecting it has something to do with that; my first hypothesis is that it might have something to do with the execute bits on one of those three files, or with the fact that `/bin/sh` should be a symlink to `busybox` and that's being messed up somewhere.

**Comment by cwoolum** [comment]
Okay, I was able to get a bash prompt. When I run `ls -la /sbin/init.resizefs`, I get

```
-rwxr-xr-x    1 0            0                2708 Jul 29  2020 /sbin/init.resizefs
```

When I try to execute it though, I get `not found`

**Comment by cwoolum** [comment]
It looks like the line endings are wrong on `init.resizefs`. I ran `file root/sbin/init.resizefs` and got back
```
root/sbin/init.resizefs: POSIX shell script, ASCII text executable, with CRLF line terminators
```

**Comment by cwoolum** [comment]
So I re-cloned this repo in WSL and re-ran the script and the line endings are correct now. It now is able to resize the volume.

I had previously cloned it in Windows and then ran it from Linux. This is a pretty wierd behavior because I would expect the files to be encoded based on the system they are executed from, not where they are cloned from.

**Comment by sgielen** [comment]
Interesting, very nice that you found the root cause! It makes sense that line endings caused this, and I wouldn't have thought of that myself.

I'd like to reopen this, because I think it should be easy to fix even when cloning natively on Windows. If I prepare a test branch, would you be willing to test if images built with that branch do work?

**Comment by sgielen** [comment]
@cwoolum I've pushed a possible fix to the `fix/dos-line-endings` branch. Would you be so kind as to check whether checking out the repo on native Windows, switching to the branch, building as you did before and then flashing & booting the image does work now?

**Comment by cwoolum** [comment]
So I did test your changes. While it does fix the line endings in the two init.* files, it does't fix the line endings on the *.sh scripts so I had to fix the line endings before I could even build the image.

I created a PR #23 that adds a `.gitattributes`. This forces windows to keep line endings for the files specified.

**Comment by cwoolum** [comment]
@sgielen, did you get a chance to look at my PR?

**Comment by sgielen** [comment]
Sorry for getting back late to this, @cwoolum.

PR #25 landed the cgroup changes, and since they were included in your PR #23 as well, it made it unmergeable. Therefore, I took your changes and committed them, giving credit to you in the log message (f1693d15099629120c3b49f1ce5db70974370d64). That should resolve your issue. If not, please reopen!
